File: src/Java/myservicebus-amazon-sqs/src/main/java/com/myservicebus/amazon/sqs/AmazonSqsSendTransport.java
package com.myservicebus.amazon.sqs;

import com.myservicebus.SendTransport;
import software.amazon.awssdk.services.sns.SnsClient;
import software.amazon.awssdk.services.sqs.SqsClient;

import java.util.Map;

public final class AmazonSqsSendTransport implements SendTransport {
    private final SqsClient sqs;
    private final SnsClient sns;
    private final AmazonSqsEndpointAddress.EntityKind kind;
    private final String destination;
    private final String entityName;

    AmazonSqsSendTransport(SqsClient sqs, SnsClient sns, AmazonSqsEndpointAddress.EntityKind kind,
            String destination, String entityName) {
        this.sqs = sqs;
        this.sns = sns;
        this.kind = kind;
        this.destination = destination;
        this.entityName = entityName;
    }

    @Override
    public void send(byte[] data, Map<String, Object> headers, String contentType) {
        try {
            if (data.length > 1_048_576) {
                throw new IllegalArgumentException("Amazon SQS/SNS messages cannot exceed 1 MiB");
            }
            if (kind == AmazonSqsEndpointAddress.EntityKind.TOPIC) {
                sns.publish(AmazonSqsMessageMapper.snsRequest(destination, data, contentType));
            } else {
                sqs.sendMessage(AmazonSqsMessageMapper.sqsRequest(destination, data, contentType));
            }
        } catch (Exception exception) {
            throw exception instanceof AmazonSqsTransportException transportException
                    ? transportException : new AmazonSqsTransportException("send", entityName, exception);
        }
    }
}

File: src/MyServiceBus.AmazonSqs/AmazonSqsSendTransport.cs
using Amazon.SimpleNotificationService;
using Amazon.SQS;
using MyServiceBus.Serialization;

namespace MyServiceBus;

public sealed class AmazonSqsSendTransport : ISendTransport
{
    private readonly IAmazonSQS _sqs;
    private readonly IAmazonSimpleNotificationService _sns;
    private readonly AmazonSqsEntityKind _kind;
    private readonly string _destination;
    private readonly string _entityName;

    internal AmazonSqsSendTransport(
        IAmazonSQS sqs,
        IAmazonSimpleNotificationService sns,
        AmazonSqsEntityKind kind,
        string destination,
        string entityName)
    {
        _sqs = sqs;
        _sns = sns;
        _kind = kind;
        _destination = destination;
        _entityName = entityName;
    }

    /// <exception cref="AmazonSqsTransportException">The AWS client rejected or could not complete the send.</exception>
    public async Task Send<T>(T message, SendContext context, CancellationToken cancellationToken = default)
        where T : class
    {
        try
        {
            var body = context.GetMessageBody(message).GetBytes();
            if (body.Length > 1_048_576)
                throw new AmazonSqsTransportException("send", _entityName,
                    new ArgumentOutOfRangeException(nameof(message), "Amazon SQS/SNS messages cannot exceed 1 MiB."));

            if (_kind == AmazonSqsEntityKind.Topic)
                await _sns.PublishAsync(AmazonSqsMessageMapper.CreateSnsRequest(
                    _destination, body, InboundMessageResolver.EnvelopeContentType), cancellationToken).ConfigureAwait(false);
            else
                await _sqs.SendMessageAsync(AmazonSqsMessageMapper.CreateSqsRequest(
                    _destination, body, InboundMessageResolver.EnvelopeContentType), cancellationToken).ConfigureAwait(false);
        }
        catch (Exception exception) when (exception is not OperationCanceledException and not AmazonSqsTransportException)
        {
            throw new AmazonSqsTransportException("send", _entityName, exception);
        }
    }
}
